oauthenticator/globus.py

```diff
@@ -0,0 +1,223 @@
+        when they start notebooks, Terminals, etc."""
+    ).tag(config=True)
+
+    def _exclude_tokens_default(self):
+        return ['auth.globus.org']
+
+    scope = List(
+        help="""Set scope for Globus Auth. The transfer scope can be removed in
+         which case a transfer token will no longer be passed to the spawner.
+         Alternatively, add additional transfer scopes and those transfer
+         tokens will automatically be added."""
+    ).tag(config=True)
+
+    def _scope_default(self):
+        return [
+            'openid',
+            'profile',
+            'urn:globus:auth:scope:transfer.api.globus.org:all'
+        ]
+
+    allow_refresh_tokens = Bool(
+        help="""Allow users to have Refresh Tokens. If Refresh Tokens are not
+        allowed, users must use regular Access Tokens which will expire after
+        a set time. Set to False for increased security, True for increased
+        convenience."""
+    ).tag(config=True)
+
+    def _allow_refresh_tokens_default(self):
+        return True
+
+    globus_local_endpoint = Unicode(help="""If Jupyterhub is also a Globus
+    endpoint, its endpoint id can be specified here.""").tag(config=True)
+
+    def _globus_local_endpoint_default(self):
+        return os.getenv('GLOBUS_LOCAL_ENDPOINT', '')
+
+    def pre_spawn_start(self, user, spawner):
+        """Add tokens to the spawner whenever the spawner starts a notebook.
+        This will allow users to create a transfer client:
+        globus-sdk-python.readthedocs.io/en/stable/tutorial/#tutorial-step4
+        """
+        if self.globus_local_endpoint:
+            spawner.environment.update(
+                {'GLOBUS_LOCAL_ENDPOINT': self.globus_local_endpoint}
+            )
+        if user.auth_state.get('globus_data'):
+            globus_data = base64.b64encode(
+                pickle.dumps(user.auth_state['globus_data'])
+            )
+            spawner.environment['GLOBUS_DATA'] = globus_data
+        else:
+            # This can happen when migrating old users with a valid
+            # Jupyterhub session that have never used this oauthenticator
+            self.log.error('Globus data not found, user will not be able '
+                           'to start transfers until they '
+                           're-authenticate.')
+
+    def globus_portal_client(self):
+        return globus_sdk.ConfidentialAppAuthClient(
+            self.client_id,
+            self.client_secret)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.globus_data = {}
+
+    @gen.coroutine
+    def authenticate(self, handler, data=None):
+        """
+        Authenticate with globus.org. Usernames (and therefore Jupyterhub
+        accounts) will correspond to a Globus User ID, so foouser@globusid.org
+        will have the 'foouser' account in Jupyterhub.
+        """
+        code = handler.get_argument("code")
```

Comment:
I went ahead and removed the error check that was removed last week, in addition to the `no_code_test` in the test code, so this should be up to date.